Introduce ReceiveAuthKey

shaavan opened the pull request:
Builds on #3845

This PR builds on the work in #3845 by introducing `ReceiveAuthKey`, a dedicated struct that replaces the previously hardcoded `[u8; 32]` used for authenticating `MessageContext`s in incoming `BlindedMessagePath`s.

It:
- Encapsulates the authentication key in a type-safe `ReceiveAuthKey` struct
- Adds this key to the `NodeSigner` interface
- Updates the `BlindedMessagePath` constructor to accept it as a parameter

This completes the original intent of #3845 — making the authentication mechanism explicit and less byte-heavy, while preserving the same security properties.

### Next steps:
- Remove the now-redundant `(HMAC, Nonce)` pair from relevant `MessageContext` variants
- Apply this authentication mechanism to dummy hops (#3726)

shaavan commented:
cc @TheBlueMatt

shaavan commented:
**Updated from** [pr3917.02](https://github.com/shaavan/rust-lightning/commits/pr3917.02) to [pr3917.03](https://github.com/shaavan/rust-lightning/commits/pr3917.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3917.02..pr3917.03)):

Changes:
1. Rebase on main to resolve merge conflicts.

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -93,7 +93,7 @@ impl BlindedMessagePath {
 				context,
 				&blinding_secret,
-				[41; 32], // TODO: Pass this in
+				ReceiveAuthKey { inner: [41; 32] }, // TODO: Pass this in
```
Thanks a lot for the pointer, and suggestion!
Fixed the commit order and history in [pr3917.04](https://github.com/shaavan/rust-lightning/commits/pr3917.04)
Thanks!

shaavan commented on lightning/src/offers/signer.rs:
```diff
@@ -38,23 +37,25 @@ const DERIVED_METADATA_AND_KEYS_HMAC_INPUT: &[u8; 16] = &[2; 16];
-const PAYMENT_HASH_HMAC_INPUT: &[u8; 16] = &[7; 16];
+// NOTE:
+// The following `HMAC_INPUT` constants were previously used for authenticating fields
+// in `OffersContext` and `AsyncPaymentsContext`, but were removed in LDK v0.1.X with
```
@TheBlueMatt

which version will this merged?

shaavan replied:
Done! [pr3917.07](https://github.com/shaavan/rust-lightning/commits/pr3917.07)
Thanks for the pointer! ✨

shaavan commented:
**Updated from** [pr3917.04](https://github.com/shaavan/rust-lightning/commits/pr3917.04) to [pr3917.05](https://github.com/shaavan/rust-lightning/commits/pr3917.05) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3917.04..pr3917.05)):

Changes:
1. Rebase on main to resolve merge conflicts.

valentinewallace commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -501,8 +507,8 @@ pub trait MessageRouter {
 	/// Creates [`BlindedMessagePath`]s to the `recipient` node. The nodes in `peers` are assumed to
 	/// be direct peers with the `recipient`.
 	fn create_blinded_paths<T: secp256k1::Signing + secp256k1::Verification>(
-		&self, recipient: PublicKey, context: MessageContext, peers: Vec<PublicKey>,
-		secp_ctx: &Secp256k1<T>,
+		&self, recipient: PublicKey, local_node_receive_key: ReceiveAuthKey,
+		context: MessageContext, peers: Vec<PublicKey>, secp_ctx: &Secp256k1<T>,
```
@TheBlueMatt I don't necessarily think this is better, but could an alternate design be: have a secret tweak for our public key, use that tweak when decoding the onion, and fall back to not tweaking our public key if that fails? It seems like that way we could just delete the fallback code after some time, we wouldn't need to add any chacha/cryptography stuff, and wouldn't have extra computation in the case where the fallback isn't needed.

valentinewallace replied:
> Alternatively, we could have the router actually just be a router - returning a path that the BlindedMessagePath should traverse and leaving it to the caller to actually build the path. This I kinda like but is definitely a larger refactor.

I do like that idea... 

I think you were following but just in case, my thought was to do something like this:
```diff
diff --git a/lightning/src/blinded_path/message.rs b/lightning/src/blinded_path/message.rs
index ba99b283e..e36d49014 100644
--- a/lightning/src/blinded_path/message.rs
+++ b/lightning/src/blinded_path/message.rs
@@ -73,7 +73,8 @@ impl BlindedMessagePath {
 	//  TODO: make all payloads the same size with padding + add dummy hops
 	pub fn new<ES: Deref, T: secp256k1::Signing + secp256k1::Verification>(
 		intermediate_nodes: &[MessageForwardNode], recipient_node_id: PublicKey,
-		context: MessageContext, entropy_source: ES, secp_ctx: &Secp256k1<T>,
+		blinded_path_auth_tweak: Option<Scalar>, context: MessageContext, entropy_source: ES,
+		secp_ctx: &Secp256k1<T>,
 	) -> Result<Self, ()>
 	where
 		ES::Target: EntropySource,
@@ -92,6 +93,7 @@ impl BlindedMessagePath {
 				secp_ctx,
 				intermediate_nodes,
 				recipient_node_id,
+				blinded_path_auth_tweak,
 				context,
 				&blinding_secret,
 			)
@@ -660,8 +662,13 @@ pub(crate) const MESSAGE_PADDING_ROUND_OFF: usize = 100;
 /// Construct blinded onion message hops for the given `intermediate_nodes` and `recipient_node_id`.
 pub(super) fn blinded_hops<T: secp256k1::Signing + secp256k1::Verification>(
 	secp_ctx: &Secp256k1<T>, intermediate_nodes: &[MessageForwardNode],
-	recipient_node_id: PublicKey, context: MessageContext, session_priv: &SecretKey,
-) -> Result<Vec<BlindedHop>, secp256k1::Error> {
+	recipient_node_id: PublicKey, blinded_path_auth_tweak: Scalar, context: MessageContext,
+	session_priv: &SecretKey,
+) -> Result<Vec<BlindedHop>, secp256k1::Error>
+where
+	NS::Target: NodeSigner,
+{
+	let recipient_node_id = recipient_node_id.mul_tweak(&secp_ctx, &blinded_path_auth_tweak)?;
 	let pks = intermediate_nodes
 		.iter()
 		.map(|node| node.node_id)
diff --git a/lightning/src/onion_message/messenger.rs b/lightning/src/onion_message/messenger.rs
index 31067fad2..bea3b3d44 100644
--- a/lightning/src/onion_message/messenger.rs
+++ b/lightning/src/onion_message/messenger.rs
@@ -1038,20 +1038,13 @@ where
 	Ok((first_node_id, message, first_node_addresses))
 }
 
-/// Decode one layer of an incoming [`OnionMessage`].
-///
-/// Returns either the next layer of the onion for forwarding or the decrypted content for the
-/// receiver.
-pub fn peel_onion_message<NS: Deref, L: Deref, CMH: Deref>(
-	msg: &OnionMessage, secp_ctx: &Secp256k1<secp256k1::All>, node_signer: NS, logger: L,
-	custom_handler: CMH,
-) -> Result<PeeledOnion<<<CMH>::Target as CustomOnionMessageHandler>::CustomMessage>, ()>
+fn onion_decode_ss<NS: Deref>(
+	node_signer: NS, blinding_point: &PublicKey, tweak: Option<Scalar>,
+) -> Result<(SharedSecret, [u8; 32]), ()>
 where
 	NS::Target: NodeSigner,
-	L::Target: Logger,
-	CMH::Target: CustomOnionMessageHandler,
 {
-	let control_tlvs_ss = match node_signer.ecdh(Recipient::Node, &msg.blinding_point, None) {
+	let control_tlvs_ss = match node_signer.ecdh(Recipient::Node, blinding_point, tweak) {
 		Ok(ss) => ss,
 		Err(e) => {
 			log_error!(logger, "Failed to retrieve node secret: {:?}", e);
@@ -1074,12 +1067,40 @@ where
 			},
 		}
 	};
-	let next_hop = onion_utils::decode_next_untagged_hop(
+	Ok((control_tlvs_ss, onion_decode_ss))
+}
+
+/// Decode one layer of an incoming [`OnionMessage`].
+///
+/// Returns either the next layer of the onion for forwarding or the decrypted content for the
+/// receiver.
+pub fn peel_onion_message<NS: Deref, L: Deref, CMH: Deref>(
+	msg: &OnionMessage, secp_ctx: &Secp256k1<secp256k1::All>, node_signer: NS, logger: L,
+	custom_handler: CMH,
+) -> Result<PeeledOnion<<<CMH>::Target as CustomOnionMessageHandler>::CustomMessage>, ()>
+where
+	NS::Target: NodeSigner,
+	L::Target: Logger,
+	CMH::Target: CustomOnionMessageHandler,
+{
+	let blinded_path_auth_tweak = node_signer.blinded_path_auth_tweak()?;
+	let (control_tlvs_ss, onion_decode_ss) = onion_decode_ss(node_signer, blinded_path_auth_tweak);
+	let mut next_hop = onion_utils::decode_next_untagged_hop(
 		onion_decode_ss,
 		&msg.onion_routing_packet.hop_data[..],
 		msg.onion_routing_packet.hmac,
 		(control_tlvs_ss, custom_handler.deref(), logger.deref()),
 	);
+	if next_hop.is_err() {
+		let (control_tlvs_ss, onion_decode_ss) = onion_decode_ss(node_signer, None);
+		next_hop = onion_utils::decode_next_untagged_hop(
+			onion_decode_ss,
+			&msg.onion_routing_packet.hop_data[..],
+			msg.onion_routing_packet.hmac,
+			(control_tlvs_ss, custom_handler.deref(), logger.deref()),
+		)
+	}
+
 	match next_hop {
 		Ok((
 			Payload::Receive {

```
Then we wouldn't need any of the dual chacha/cryptography stuff or to modify the blinded path utils, and could just delete the fallback code above after some time. It just seems simpler code-wise. But we'd still need to pass the tweak into the router as you say.

valentinewallace reviewed (commented):
Finished a pass, looks solid! Still need to review the fuzzing code moves in detail. Will take another pass also looking closely at the dual read adapter again next week!

valentinewallace commented on lightning/src/crypto/chacha20poly1305rfc.rs:
```diff
@@ -10,247 +10,148 @@
+		let zero_key = [0u8; 64];
+		cipher.process(&zero_key, &mut mac_key);
+
+		#[cfg(not(fuzzing))]
+		let mut mac = Poly1305::new(&mac_key[..32]);
+		#[cfg(fuzzing)]
+		let mut mac = Poly1305::new(&key);
```
I can also do this in follow-up. Not a huge deal but would prefer to either do this or at least have a doc comment on this fuzzing-gated code.

valentinewallace replied:
Hmm, I guess that also already happens whenever a new feerate call is added? The diff above is a more significant change to the hardcoded strings though. Adding a comment seems fine too.

shaavan replied:
Got it! Since it’s just a fuzzing-only quirk, we can probably address it in a follow-up if it ever turns into an issue.
Thanks so much for the clarification! ✨

valentinewallace commented on lightning/src/blinded_path/utils.rs:
```diff
@@ -193,9 +201,19 @@ where
+	if let Some(hop_recv_key) = hop_recv_key {
+		chachapoly_encrypt_with_swapped_aad(payload_data, encrypted_tlvs_rho, hop_recv_key.0)
+	} else {
+		let mut chacha = ChaCha20Poly1305RFC::new(&encrypted_tlvs_rho, &[0; 12], &[]);
+		let mut tag = [0; 16];
+		chacha.encrypt_full_message_in_place(&mut payload_data, &mut tag);
+		payload_data.extend_from_slice(&tag);
+		payload_data
```
It looks like this can be reverted to the simpler version of:
```rust
	let write_adapter = ChaChaPolyWriteAdapter::new(encrypted_tlvs_rho, &payload);
	write_adapter.encode()

shaavan replied:
Ahh, makes sense! I checked and it looks like the logic is identical!
Simplified in [pr3917.07](https://github.com/shaavan/rust-lightning/commits/pr3917.07). Thanks!

valentinewallace reviewed (commented):
LGTM after outstanding feedback. Let me know if you don't have time for this @shaavan, I'm going to be OOO next week so would like to get this in

shaavan commented:
**Updated from** [pr3917.06](https://github.com/shaavan/rust-lightning/commits/pr3917.06) to [pr3917.07](https://github.com/shaavan/rust-lightning/commits/pr3917.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3917.06..pr3917.07)):
Addressed @TheBlueMatt, @valentinewallace comments

Changes:
1. Various documentation and code clean-ups.
2. Introduce usage of `control_tlvs_authenticated` in code.

shaavan commented on lightning/src/offers/signer.rs:
```diff
@@ -38,23 +37,25 @@ const DERIVED_METADATA_AND_KEYS_HMAC_INPUT: &[u8; 16] = &[2; 16];
+// Reserved HMAC_INPUT values — do not reuse:
+//
+// const OFFER_PAYMENT_ID_HMAC_INPUT: &[u8; 16] = &[5; 16];
+// #[cfg(async_payments)]
```
Got it! Removed the comment for async_payments specific field in [pr3917.07](https://github.com/shaavan/rust-lightning/commits/pr3917.07)
Thanks!

shaavan commented on lightning/src/crypto/poly1305.rs:
```diff
@@ -7,386 +7,426 @@
-
-		poly
+#[cfg(not(fuzzing))]
+mod fuzzy_poly1305 {
```
Thanks for the suggestion and the diff! ✨
Cleaned up in [pr3917.07](https://github.com/shaavan/rust-lightning/commits/pr3917.07)

shaavan commented on lightning/src/blinded_path/utils.rs:
```diff
@@ -145,14 +149,16 @@ where
 
-	for pk in unblinded_path {
+	let mut iter = unblinded_path.peekable();
+	while let Some(pk) = iter.next() {
```
Makes sense! Cleaned up in [pr3917.07](https://github.com/shaavan/rust-lightning/commits/pr3917.07)
Thanks!

shaavan commented on lightning/src/crypto/streams.rs:
```diff
@@ -49,6 +53,132 @@ impl<'a, T: Writeable> Writeable for ChaChaPolyWriteAdapter<'a, T> {
+impl<'a, R: Read> Read for ChaChaDualPolyReader<'a, R> {
+	// Decrypt bytes from Self::read into `dest`.
+	// `ChaCha20Poly1305RFC::finish_and_check_tag` must be called to check the tag after all reads
+	// complete.
```
Thanks for catching it! Fixed in [pr3917.07](https://github.com/shaavan/rust-lightning/commits/pr3917.07)

valentinewallace commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1074,22 +1074,29 @@ where
 			None,
 		)) => match (message, context) {
 			(ParsedOnionMessageContents::Offers(msg), Some(MessageContext::Offers(ctx))) => {
+				// Note: We introduced the `control_tlvs_authenticated` check in LDK v0.2
+				// to simplify and standardize onion message authentication.
+				// To continue supporting offers/refunds created before v0.2, we allow
+				// unauthenticated control TLVs for these messages, as they were verified
+				// using the legacy method.
 				Ok(PeeledOnion::Offers(msg, Some(ctx), reply_path))
```
I believe we should also be failing on unauthenticated TLVs for some `OffersContext` variants but not others, i.e. succeed if the blinded path was within a legacy offer but not if the blinded path is a reply path to an invreq or invoice. I think I prefer to also drop support for legacy refunds to simplify the code:
```diff
diff --git a/lightning/src/onion_message/messenger.rs b/lightning/src/onion_message/messenger.rs
index 0cb6e3f09..60bee7fa9 100644
--- a/lightning/src/onion_message/messenger.rs
+++ b/lightning/src/onion_message/messenger.rs
@@ -1119,11 +1119,19 @@ where
                        None,
                )) => match (message, context) {
                        (ParsedOnionMessageContents::Offers(msg), Some(MessageContext::Offers(ctx))) => {
-                               // Note: We introduced the `control_tlvs_authenticated` check in LDK v0.2
-                               // to simplify and standardize onion message authentication.
-                               // To continue supporting offers/refunds created before v0.2, we allow
-                               // unauthenticated control TLVs for these messages, as they were verified
-                               // using the legacy method.
+                               match ctx {
+                                       OffersContext::InvoiceRequest { .. } => {
+                                               // Note: We introduced the `control_tlvs_authenticated` check in LDK v0.2 to simplify
+                                               // and standardize onion message authentication.
+                                               // To continue supporting offers created before v0.2, we allow unauthenticated control
+                                               // TLVs for these messages, as they were verified using the legacy method.
+                                       },
+                                       _ if !control_tlvs_authenticated => {
+                                               log_trace!(logger, "Received an unauthenticated offers onion message");
+                                               return Err(());
+                                       },
+                                       _ => {},
+                               }
                                Ok(PeeledOnion::Offers(msg, Some(ctx), reply_path))
                        },
                        (ParsedOnionMessageContents::Offers(msg), None) => {

shaavan replied:
Makes sense! Since all offer message responses are short-lived by nature, and refunds always include an absolute expiry, it’s makes sense to drop legacy support for them while retaining it for long-lived offers.
If I understand correctly, we’ll also remove this legacy support once v0.2 is widely adopted. (Let me know just in case if I got it wrong! 😄)

I’ve updated the logic in [pr3917.08](https://github.com/shaavan/rust-lightning/commits/pr3917.08). Thanks!

valentinewallace replied:
Hm, that seems like a larger change if we're updating all the `OffersMessageHandler`/`AsyncPaymentsMessageHandler`, etc trait methods. It'll be mechanical changes aside from getting the docs to be clear but still. Do you have any thoughts about how to go about this? I feel like doing it in follow-up would be the way to go.

valentinewallace replied:
Discussed offline, kept the `control_tlvs_authenticated` check in the messenger but added support for legacy refunds. Reasoning is that we don't want to support receiving to blinded paths that we didn't create, and we can distinguish that case from receiving to our node_id directly based on whether the provided message context is `Some`, so no need to update all the handlers.

valentinewallace reviewed (approved):
Just have one thing that can be punted if we want to land. Thank you @shaavan!

shaavan commented:
**Updated from** [pr3917.07](https://github.com/shaavan/rust-lightning/commits/pr3917.07) to [pr3917.08](https://github.com/shaavan/rust-lightning/commits/pr3917.08) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3917.07..pr3917.08)):
Addressed @valentinewallace comments

Changes:
1. Renamed `mod poly1305` to fix linting ci
1. Updated `peel_onion_message` logic.

valentinewallace commented on lightning/src/crypto/streams.rs:
```diff
@@ -49,6 +53,132 @@ impl<'a, T: Writeable> Writeable for ChaChaPolyWriteAdapter<'a, T> {
+			ChaChaDualPolyReader { chacha: &mut chacha, poly: &mut mac, read_len: 0, read: s };
+
+		let readable: T = Readable::read(&mut chacha_stream)?;
+		chacha_stream.read.eat_remaining()?;
```
Shouldn't we just error if any bytes remain, since it means the length we were given for the data was wrong?

valentinewallace commented:
Removed legacy refund support and some references to async payments cfg-gated code. [Diff](https://github.com/lightningdevkit/rust-lightning/compare/f1f3c2f7aac7e90a12892e7b4c8889427b86a092..d3120ba065a8f9a8c8a727974601d9c7e5a22c0f)

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -13942,38 +13883,26 @@ where
-								payment_id, PaymentFailureReason::InvoiceRequestRejected,
-							);
-						}
+					Some(OffersContext::OutboundPayment { payment_id, .. }) => {
```
Ugh, good catch. Removed legacy refund support.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3761,7 +3707,7 @@ where
 			ChainHash::using_genesis_block(params.network), params.best_block,
 			our_network_pubkey, current_timestamp, expanded_inbound_key,
-			secp_ctx.clone(), message_router
+			node_signer.get_receive_auth_key(), secp_ctx.clone(), message_router
```
Weird, this method is `#[rustfmt::skip]`'d. Fixed.
